crates/sui-indexer/src/fetcher/transaction_handler.rs

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use sui_json_rpc_types::TransactionsPage;
+use sui_sdk::SuiClient;
+use sui_types::base_types::TransactionDigest;
+use sui_types::query::TransactionQuery;
+use tracing::{info, warn};
+
+// TODO: want to tune the size
+const TRANSACTION_PAGE_SIZE: usize = 100;
+
+pub struct TransactionHandler {
+    rpc_client: SuiClient,
+}
+
+impl TransactionHandler {
+    pub fn new(rpc_client: SuiClient) -> Self {
+        Self { rpc_client }
+    }
+
+    pub async fn run_forever(&self) {
+        // TODO: read next cursor from DB if available
+        let mut next_cursor = None;
+        loop {
+            let page = self.get_transaction_page(next_cursor).await;
+            if let Ok(page) = page {
+                info!("Current transaction page {:?}", page.data);
+                // TODO: write this page of txns to DB
```

Comment:
We should separate transaction fetching and handling(i.e., saving to db)

```

trait TransactionFetcher {
 async getNextBatch(cursor: String);
}

trait TransactionHandler {
 async handleBatch(txns: Transactions[]);
}

```
That way you can have`RpcTransactionFetcher` implementing `TransactionFetcher`, and maybe at some point, we can have also have a`EmbeddedTransactionFetcher`(similar to how we implement rosetta as a [augmented fullnode](https://github.com/MystenLabs/sui/blob/f6c272b9a79cce7d27693472e6fbce6caba974ab/crates/sui-rosetta/src/main.rs#L119)) implementing the same trait.

Similarly, we can then also have `SaveToPostgresHandler` implementing `TransactionHandler`

Reply:
synced with @666lcz on this, overall I agree that it will be better that way and will do the splitting in a following PR.